### PR TITLE
[Bug][Kotlin-client] Can now handle path param of type list

### LINKED ISSUE #7199: [BUG][kotlin] Path params with type: array doesn't work
##### Description

If API contains path parameters of type array, generated client function uses toString() method on kotlin.Array which doesn't provide comma separated representation.

Generated code:
```
fun <function-name-generated-resp-operationId>(<my-array-param>: kotlin.Array<kotlin.String>) : kotlin.Array<MyDtoType> {
        val localVariableBody: kotlin.Any? = null
        val localVariableQuery: MultiValueMap = mutableMapOf()
        val localVariableHeaders: MutableMap<String, String> = mutableMapOf()
        val localVariableConfig = RequestConfig(
            RequestMethod.GET,
            "/path-part-1/{<my-array-param>}/path-part-2".replace("{"+"<my-array-param>"+"}", "$<my-array-param>"),
            query = localVariableQuery,
            headers = localVariableHeaders
        )
```

The problem is the string templating here (`"$<my-array-param>"`), which uses objects toString() method.

##### openapi-generator version

4.3.1

##### OpenAPI declaration file content or url

Path param of type array:
```
parameters:
        - name: <my-array-param>
          in: path
          required: true
          style: simple
          explode: false
          schema:
            type: array
            items:
              type: string
```

##### Generation Details

We are using openapi-generator-maven-plugin.

##### Steps to reproduce

Every API with a path parameter of type array should provide the described behaviour.

##### Suggest a fix

Use joinToString method:
```
"${<my-array-param>.joinToString(",")}"
```

### PR DESCRIPTION
**Pull request by sjoblomj**
In #11228 I fixed #7199. However, the fix was only for library okhttp, and it turns out that also volley and multiplatform suffer from the same problem. This PR fixes this bug.

The PR also contains some cleaning up of the generated code (which is what most of the updated sample files is about).

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@jimschubert @dr4ke616 @karismann @Zomzog @andrewemery @4brunu @yutaka0m @wing328

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/api.mustache [hunk 1]
```diff
@@ -52,47 +52,34 @@ class {{classname}} (
         {{^bodyParam}}
         val body: Any? = null
         {{/bodyParam}}
-        {{#allParams}}
-        {{#required}}
-        // verify the required parameter '{{paramName}}' is set
-        // This is probably taken care of by non-null types anyway
-        requireNotNull({{paramName}})
-        {{/required}}
-        {{/allParams}}
```
Comment (sjoblomj) [inline modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/api.mustache:61]: The Kotlin type systems makes sure that the parameters cannot be null. Completely redundant check.

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/gradle.properties.mustache [hunk 2]
```diff
@@ -1,4 +1,4 @@
-{{#generateRoomModels}}
 android.useAndroidX=true
+{{#generateRoomModels}}
```
Comment (sjoblomj) [inline modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/gradle.properties.mustache:2]: jvm-volley refuses to build without `android.useAndroidX=true`, even if generateRoomModels is false

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-retrofit2/infrastructure/ApiClient.kt.mustache [hunk 3]
```diff
@@ -100,13 +100,9 @@ import okhttp3.MediaType.Companion.toMediaType
...
-            .addInterceptor(HttpLoggingInterceptor(object : HttpLoggingInterceptor.Logger {
-                override fun log(message: String) {
-                    logger?.invoke(message)
-                }
-            }).apply {
-                level = HttpLoggingInterceptor.Level.BODY
-            })
+            .addInterceptor(HttpLoggingInterceptor { message -> logger?.invoke(message) }
+                .apply { level = HttpLoggingInterceptor.Level.BODY }
+            )
```
Comment (sjoblomj) [inline modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-retrofit2/infrastructure/ApiClient.kt.mustache:105]: Simplifying code

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/api.mustache [hunk 4]
```diff
@@ -52,47 +52,34 @@ class {{classname}} (
...
 
-        // Do some work or avoid some work based on what we know about the model, 
+        // Do some work or avoid some work based on what we know about the model,
         // before we delegate to a pluggable request factory template
         // The request factory template contains only pure code and no templates
         // to make it easy to override with your own.
 
         // create path and map variables
-        val path = "{{{path}}}"{{#pathParams}}.replace("{" + "{{baseName}}" + "}", IRequestFactory.escapeString({{{paramName}}}.toString())){{/pathParams}};
+        val path = "{{{path}}}"{{#pathParams}}.replace("{" + "{{baseName}}" + "}", {{#isContainer}}{{paramName}}.joinToString(","){{/isContainer}}{{^isContainer}}IRequestFactory.escapeString({{{paramName}}}.toString()){{/isContainer}}){{/pathParams}}
```
Comment (sjoblomj) [inline modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/api.mustache:65]: Bugfix (#7199) to handle that you can't use `toString()` on lists

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/api.mustache [hunk 5]
```diff
@@ -60,19 +60,14 @@ import kotlinx.serialization.encoding.*
...
-        {{/headerParams}}
+        val localVariableQuery = mutableMapOf<String, List<String>>(){{#queryParams}}
+        {{{paramName}}}?.apply { localVariableQuery["{{baseName}}"] = {{#isContainer}}toMultiValue(this, "{{collectionFormat}}"){{/isContainer}}{{^isContainer}}listOf("${{{paramName}}}"){{/isContainer}} }{{/queryParams}}
+        val localVariableHeaders = mutableMapOf<String, String>(){{#headerParams}}
+        {{{paramName}}}?.apply { localVariableHeaders["{{baseName}}"] = {{#isContainer}}this.joinToString(separator = collectionDelimiter("{{collectionFormat}}")){{/isContainer}}{{^isContainer}}this.toString(){{/isContainer}} }{{/headerParams}}
 
         val localVariableConfig = RequestConfig<kotlin.Any?>(
             RequestMethod.{{httpMethod}},
-            "{{path}}"{{#pathParams}}.replace("{" + "{{baseName}}" + "}", "${{{paramName}}}"){{/pathParams}},
+            "{{path}}"{{#pathParams}}.replace("{" + "{{baseName}}" + "}", {{#isContainer}}{{paramName}}.joinToString(","){{/isContainer}}{{^isContainer}}"${{{paramName}}}"{{/isContainer}}){{/pathParams}},
```
Comment (sjoblomj) [inline modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/api.mustache:70]: Bugfix (#7199) to handle that you can't use `toString()` on lists

**Comment by wing328** [comment]
LGTM. Thanks for the fix.

**Pull request merged**